pkg/runner/runner.go

```diff
@@ -0,0 +1,222 @@
+package runner
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/GoogleContainerTools/kpt/e2e"
+	"github.com/GoogleContainerTools/kpt/run"
+)
+
+// Runner runs an e2e test
+type Runner struct {
+	pkgName  string
+	testCase TestCase
+}
+
+const (
+	expectedDir         string = ".expected"
+	expectedResultsFile string = "results.yaml"
+	expectedDiffFile    string = "diff.patch"
+	expectedConfigFile  string = "config.yaml"
+)
+
+// NewRunner returns a new runner for pkg
+func NewRunner(testCase TestCase) (*Runner, error) {
+	info, err := os.Stat(testCase.Path)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open path %s: %w", testCase.Path, err)
+	}
+	if !info.IsDir() {
+		return nil, fmt.Errorf("path %s is not a directory", testCase.Path)
+	}
+	return &Runner{
+		pkgName:  filepath.Base(testCase.Path),
+		testCase: testCase,
+	}, nil
+}
+
+// Run runs the test.
+func (r *Runner) Run(cmd string, args []string, t *testing.T) error {
+	fmt.Printf("Running test against package %s\n", r.pkgName)
+	tmpDir, err := ioutil.TempDir("", "kpt-fn-catalog-e2e-*")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary dir: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpPkgPath := filepath.Join(tmpDir, r.pkgName)
+	// create dir to store untouched pkg to compare against
+	orgPkgPath := filepath.Join(tmpDir, "results")
```

Comment:
The `results` dir is not used to store original package content. It's used to hold the kpt output when it has error. Is this changed in pipeline?

Reply:
I thought results dir was meant to originally meant for the functions to write their result to that dir. 
https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/127a1afe906e8b2a7d1d37fc14106b03d9631ca8/tests/e2etest/runner/runner.go#L70

In the original harness that you wrote, my understanding is that it runs kpt fn run on tmpPkgPath but stores the results in results dir and then compares them. 

kpt pipeline does not have a --results-dir option. The idea I had was to alter the harness so that it saves the original package in orgPkgPath, runs kpt <cmd> run on tmpPkgPath, and then compares them. 

If this makes sense I can change the comment and the name of the dir to make it more clear, but please correct me if I am misunderstanding something

Reply:
in pipeline there is no --results-dir argument so I changed it to transform tmpPkgPath and compare it against orgPkgPath. but it is different and that doesn't work for fn so I updated the PR to handle fn differently